merge `vscode-python` upstream 2024.14.1

isabelizimm opened the pull request:
This applies changes from [v2024.12.0](https://github.com/microsoft/vscode-python/releases/tag/v2024.12.0), [v2024.12.1](https://github.com/microsoft/vscode-python/releases/tag/v2024.12.1), [v2024.12.2](https://github.com/microsoft/vscode-python/releases/tag/v2024.12.2), [v2024.12.3](https://github.com/microsoft/vscode-python/releases/tag/v2024.12.3), [v2024.14.0](https://github.com/microsoft/vscode-python/releases/tag/v2024.14.0), [v2024.14.1](https://github.com/microsoft/vscode-python/releases/tag/v2024.14.1).

- The majority of work for this merge is around the Native REPL, which we don't use at all.
- Move from to typescript 4.5.5 to ~5.2
- The [Rust env finder](https://github.com/microsoft/python-environment-tools) is the main locator upstream now. We currently don't use it, but may want to take steps for that change sooner rather than later. All things will work normally still as the builtin locator is still supported.

positron-bot[bot] committed:
apply changes from upstream 2024.10-2024.14.1

testlabauto committed:
Fix overlapping test result file paths (#4838)

The electron and browser tests were using the same results path and thus
only web tests were being report to TestRail.

Also added a run summary block for the web tests.

### QA Notes

Should see full electron run reported to testrail.

isabelizimm committed:
update polars

isabelizimm committed:
prePublish fixes for assert and chai

positron-bot[bot] committed:
move from fsapi to fs

isabelizimm committed:
bump numpy for  solely python 3.9

isabelizimm committed:
smartsend stuff has been renamed

isabelizimm committed:
startREPL->startTerminalREPL

isabelizimm committed:
pin torch

isabelizimm committed:
pin torch again

isabelizimm committed:
rearrange imports

isabelizimm committed:
allow all polars

sharon-wang committed:
Fix Python help url resolution on Server Web and Workbench (#4837)

### Description
- addresses part of https://github.com/posit-dev/positron/issues/4274
- a follow-up to https://github.com/posit-dev/positron/pull/4814 to fix
Python help docs in Positron Server and Positron on Workbench

#### Screenshots of the issue

<img width="865" alt="image"
src="https://github.com/user-attachments/assets/631488c4-cee6-49ec-8c9a-314830c925e7">


![image](https://github.com/user-attachments/assets/b0594155-f357-46e6-8a61-4a3ac211f600)

#### What it looks like now!

<img width="1727" alt="image"
src="https://github.com/user-attachments/assets/313a33a1-4698-47ac-8df2-314a6cc034df">

#### Implementation Notes
- adds back in the `sourceUrl.pathname` fix that I should have had in
#4814, but accidentally deleted
- appends the search query to the target path, which was getting dropped
in the proxied request in `src/vs/server/node/webClientServer.ts`

Note: the change to `src/vs/server/node/webClientServer.ts` needs to be
contributed to our upstream.

### QA Notes
In Positron Server Web and Positron on Workbench:

1. Select a Python interpreter
2. In the Console, get help for something, e.g. `import time` then
`?time` and hit Enter
3. What to expect in the Help Pane:
- the pane should display the help content with styling that matches the
current IDE theme
    - links to other help pages should load and work when clicked
- there shouldn't be 404 Not Found errors in the Developer Console or
Network tab for paths referenced in the help HTML

---------

Co-authored-by: Brian Lambert <brianlambert@gmail.com>
Co-authored-by: Jonathan <jonathan@rstudio.com>

DavisVaughan committed:
Ensure `addActivityItemStream()` doesn't lose `Output` vs `Error` stream specialization (#4848)

Addresses https://github.com/posit-dev/positron/issues/4519

In `addActivityItemStream()`, either a `ActivityItemOutputStream` or
`ActivityItemErrorStream` always goes in, but it was possible for a
"generic" `ActivityItemStream` to come out here when we make one from
the `remainderText` after the last newline:


https://github.com/posit-dev/positron/blob/e51332d5baa59b72f5e8b28861bbb751f25452ea/src/vs/workbench/services/positronConsole/browser/classes/activityItemStream.ts#L118-L136

That generic `ActivityItemStream` would get set here:

https://github.com/posit-dev/positron/blob/c9504d307c686ce2deefdc6887ad2f7b044dd2e9/src/vs/workbench/services/positronConsole/browser/classes/runtimeItemActivity.ts#L110-L116

And then pushed onto `this._activityItems`

https://github.com/posit-dev/positron/blob/c9504d307c686ce2deefdc6887ad2f7b044dd2e9/src/vs/workbench/services/positronConsole/browser/classes/runtimeItemActivity.ts#L138-L139

The problem is that our Console code doesn't know how to deal with a
generic `ActivityItemStream`. It must be a specialized
`ActivityItemOutputStream` or `ActivityItemErrorStream`!

https://github.com/posit-dev/positron/blob/c9504d307c686ce2deefdc6887ad2f7b044dd2e9/src/vs/workbench/contrib/positronConsole/browser/components/runtimeActivity.tsx#L46-L49

If a generic `ActivityItemStream` slipped through, the output would
never appear in the Console, making it look "dropped", as reported in
https://github.com/posit-dev/positron/issues/4519.

---

I like how we have a generic `ActivityItemStream` class that we
specialize for the output/error types, so I've fixed this bug by using
[polymorphic
`this`](https://www.typescriptlang.org/docs/handbook/advanced-types.html#polymorphic-this-types)
to allow `addActivityItemStream()` to take and return objects of the
correct specialized type.

Ideally we'd simply be able to use `new this.constructor()` in place of
`new ActivityItemStream()` to use the polymorphic constructor when we
need to create new objects, but due to some [well known
weirdness](https://github.com/microsoft/TypeScript/issues/3841) in
typescript, you have to manually cast it to the constructor type, so
`newActivityItemStream()` wraps that up.

I've also made the `ActivityItemStream` constructor `protected` to avoid
us using it by accident in the codebase. We should _always_ be creating
`ActivityItemOutputStream` or `ActivityItemErrorStream` instead.

Before 😢 



https://github.com/user-attachments/assets/b1854d9e-94eb-4952-aead-bebcf5c5d6a2



After 🥳 


https://github.com/user-attachments/assets/4a76974c-6862-4e84-a296-620717ddb8da

isabelizimm committed:
polars for 3.8-3.12

isabelizimm committed:
add startNativeREPL title

isabelizimm committed:
update eslint for linting errors

GCRev committed:
Improve keyboard navigation for data viewer filter dialogs (#4620)

- If focus is on the column drop down, then pressing a key will
immediately open the search sub-dialog and use that key as an initial
search input.
- Updating the search string will always place the cursor on the first
available item
- Hitting "ArrowDown" or "Tab" from within the search input will shift
focus to the list and allow keyboard navigation as expected
- Hitting "Enter" or "Space" when navigating the list will confirm the
selection

I've also added a new style to the filter widgets to use monospace font
for the values. This makes them more legible in my opinion, particularly
with values which contain "l" "1" and "I".

isabelizimm committed:
fixing tests

isabelizimm committed:
lint

isabelizimm committed:
format fixing

juliasilge committed:
Update logs in the `jupyter-adapter` extension (#4845)

Addresses #4275 

This PR makes changes to all the output channels generated by the
`jupyter-adapter` extension.

The "main" Jupyter Adapter output channel is now a real log, with
timestamps and log level annotations:

![Screenshot 2024-09-28 at 4 51
07 PM](https://github.com/user-attachments/assets/28c48414-f7ee-4da5-9ebb-eeaa96d892ac)

The existing "Console: X.Y.Z" channel now only has messages generated
from Positron, not the kernel itself. This is now also a real log and is
the part that directly addresses #4275:

![Screenshot 2024-09-28 at 4 51
26 PM](https://github.com/user-attachments/assets/9e1b1345-4001-44d6-b43a-d2b9e092027c)

Notice the new "error" log level here; you can now set the log level in
these first two output channels and it will update moving forward.

These first two output channels are generated in the same way most of
the time (we write to the "Console" channel if that kernel is up but to
the "Jupyter Adapter" channel if not) so we have to make changes to them
together; this seems good rather than bad to me and overall both are now
improved.

There is a new "Kernel: X.Y.Z" channel which still uses the raw output
channel that we created back in #1589. It is not a real log and does not
have timestamps, unless the kernel log generated those. You cannot
change the log level via the normal method for this one:

![Screenshot 2024-09-28 at 4 51
38 PM](https://github.com/user-attachments/assets/2df12fb3-7cc5-4aab-b49b-933fc3ecdc7a)

It is kind of a bummer to now have _three_ output channels generated
from this one extension but IMO the tradeoffs are worth it.


### QA Notes

You'll now see timestamps and log level annotations in the "Jupyter
Adapter" and "Console: X.Y.Z" output channels, and there is a new
"Kernel: X.Y.Z" channel.

isabelizimm committed:
remove setup for createTypeMoq

isabelizimm committed:
mock vscode.workspace since it is non-config non-writable

isabelizimm committed:
rearrange for WebAppCommands

isabelizimm committed:
scale back ruff

isabelizimm commented on extensions/positron-python/src/test/mocks/helper.ts:
```diff
@@ -1,11 +1,30 @@
+    // is the problem. We still need to make it thenable though. See this issue: https://github.com/florinn/typemoq/issues/67
+    const result = TypeMoq.Mock.ofType<T>(targetCtor, behavior, shouldOverrideTarget, ...targetCtorArgs);
+    // --- Start Positron ---
+    // result.setup((x: any) => x.then).returns(() => undefined);
```
this line caused every single usage of this function to fail with `TypeError: Cannot convert a Symbol value to a string`-- I'm not sure why since it seems to be a community vetted way to mock things resolved as promises and works upstream 😩

sharon-wang replied:
Looks like we're running into a version of https://github.com/inversify/InversifyJS/issues/1510 based on the error output, but it is odd there seemingly aren't issues with it upstream 🤷 

```
1) Interpreters service
       Changes to active document should invoke interpreter.refresh method:
     TypeError: Cannot convert a Symbol value to a string
      at Object.LAZY_IN_SYNC (node_modules/inversify/lib/constants/error_msgs.js:25:80)
      at Container._getButThrowIfAsync (node_modules/inversify/lib/container/container.js:601:40)
      at Container.get (node_modules/inversify/lib/container/container.js:374:21)
      at ServiceContainer.get (out/client/ioc/container.js:27:89)
      at new InterpreterService (out/client/interpreter/interpreterService.js:86:52)
      at Context.<anonymous> (out/test/interpreters/interpreterService.unit.test.js:142:25)
      at process.processImmediate (node:internal/timers:478:21)
      at process.callbackTrampoline (node:internal/async_hooks:130:17)
```

seeM replied:
I ran into this when we upgraded to TS v5 previously. We have some unfortunate hacks in place that helped back then. See https://github.com/posit-dev/positron/issues/2271#issuecomment-1952404450.

seeM replied:
Ah, I just noticed that they've also upgraded to TS v5 and have implemented their own solution to the above problem. We should probably revert our own workaround in `src/test/vscode-mock.ts` (original PR: https://github.com/posit-dev/positron-python/pull/365). I'm happy to look into this!

isabelizimm replied:
Yeah, a lot of this update is the TS 4.5.5 -> 5.2 updates. I'll open up a separate issue to revert the workaround and assign it to you if that works!

isabelizimm commented on extensions/positron-python/src/test/positron/webAppCommands.unit.test.ts:
```diff
@@ -59,17 +59,19 @@ suite('Web app commands', () => {
-                return path.relative(workspacePath, fsPath);
-            },
-        }));
+        // Create a mock object for `asRelativePath`
+        const mockAsRelativePath = sinon.mock().callsFake((pathOrUri: string | vscode.Uri) => {
+            const fsPath = typeof pathOrUri === 'string' ? pathOrUri : pathOrUri.fsPath;
+            return path.relative(workspacePath, fsPath);
+        });
+        // TODO @isabelizimm: HACK HACK HACK
+        // Replace the original `asRelativePath` method with the mock object
+        vscode.workspace.asRelativePath = mockAsRelativePath;
```
This was returning `TypeError: Descriptor for property workspace is non-configurable and non-writable` but if you stub the property `asRelativePath` instead of `workspace`, I hit `TypeError: Cannot stub non-existent property asRelativePath` instead. My workaround is to mock and then patch this

sharon-wang replied:
A nit style suggestion to match the stubbing style above:

```suggestion
        // Stub `vscode.workspace.asRelativePath`.
        vscode.workspace.asRelativePath = ((pathOrUri: string | vscode.Uri) => {
            const fsPath = typeof pathOrUri === 'string' ? pathOrUri : pathOrUri.fsPath;
            return path.relative(workspacePath, fsPath);
        });
```

isabelizimm committed:
turn off iterEnvs() monterey test

isabelizimm commented:
I played around with things very briefly, but if others can kick the tires, that would be much appreciated!

dfalbel committed:
Data explorer: Make date summary stats optional (#4564)

Allows for summary stats for dates to be optional, so we make the
behavior similar to the behavior of number statistics.

jonvanausdeln committed:
Windows CI Smoke Tests Part 2 (#4857)

### Intent

Continue adding more tests to windows nighlty CI. Next PR will add slack
& testrail reporting.

### Approach
Due to a very small application window size, some changes to the tests
had to be made

* Added dataexplorer tests
  - Had to hide both the console & primary (left) sidebar
  - 100x100 worked without modification
* Notebook creation test
  - Needed `notebook` layout to give window enough space

### QA
All tests pass on Linux & windows CI

testlabauto committed:
More web test cases (#4852)

Enabling more web smoke tests.

Remaining tests that are not enabled for web are commented with issue
number or note.

Also, added better handling for console tests and conditional (on web)
turn off of auto save for top action bar tests.

### QA Notes

All smoke tests pass.

jmcphers committed:
Bump version to 2024.10.0 (#4861)

isabelizimm commented on extensions/positron-python/.github/workflows/build.yml:
```diff
@@ -254,6 +254,13 @@ jobs:
       - name: Install test requirements
         run: python -m pip install --upgrade -r build/test-requirements.txt
 
+      - name: Rust Tool Chain setup
```
I don't think this should affect our build, but noting in case we should comment it out until we use the `python-environment-tools` Python locator

sharon-wang replied:
Yeah I _think_ we don't use the workflows in this directory, so this is should be okay as-is!

juliasilge committed:
Bump minimum renv version to 1.0.9 (#4854)

Addresses #3439 

### QA Notes

If you have renv < 1.0.8 installed and trigger the `"r.renvInit"`
command (via the New Project Wizard or directly) you will see:

![Screenshot 2024-09-30 at 3 28
20 PM](https://github.com/user-attachments/assets/3ae90841-c765-4abf-bded-a79e26be26f5)

isabelizimm committed:
[python] support shiny apps if ext not installed (#4786)

<!-- Thank you for submitting a pull request.
If this is your first pull request you can find information about
contributing here:
  * https://github.com/posit-dev/positron/blob/main/CONTRIBUTING.md

We recommend synchronizing your branch with the latest changes in the
main branch by either pulling or rebasing.
-->

<!--
  Describe briefly what problem this pull request resolves, or what
  new feature it introduces. Include screenshots of any new or altered
  UI. Link to any GitHub issues but avoid "magic" keywords that will 
  automatically close the issue. If there are any details about your 
approach that are unintuitive or you want to draw attention to, please
  describe them here.
-->

related to #4675 

### QA Notes

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

create file named `app.py`
```python
from shiny import App, reactive, render, ui

app_ui = ui.page_fluid(
    ui.input_action_button("action_button", "Action"),  
    ui.output_text("counter"),
)

def server(input, output, session):
    @render.text()
    @reactive.event(input.action_button)
    def counter():
        return f"{input.action_button()}"

app = App(app_ui, server)

```

when shiny ext installed:

![Screenshot 2024-09-30 at 11 49
44 AM](https://github.com/user-attachments/assets/09f20c88-f255-4801-a918-4452e66edd33)

when shiny ext not installed/enabled, should have `Run Shiny App in
Terminal` as first button

sharon-wang committed:
Fix interactive R and Python plots on Server Web and Workbench (#4855)

### Description

- part of #4274
- addresses https://github.com/posit-dev/positron/issues/4804
- should unblock #4806, where the new plotly rendering method relies on
a proxy server instead of an HTML file proxy

#### Changes

- add a new command to create a generic proxy server
`positronProxy.startHttpProxyServer`
- rename the command `positronProxy.stopHelpProxyServer` to
`positronProxy.stopProxyServer` since it is not help-specific
- rename `resources/scripts.html` to `resources/scripts_help.html` since
it is help-specific
- move the src/href rewriting to a private reusable function
`rewriteUrlsWithProxyPath`, which is now used by the generic http proxy
and the help proxy `contentRewriter`
- update `src/vs/code/browser/workbench/workbench.ts` to resolve the uri
while maintaining the uri's original path, query string and fragment
strings (NOTE: needs to be contributed upstream)
- update
`src/vs/workbench/services/languageRuntime/common/languageRuntimeUiClient.ts`
to choose between starting an HTML file proxy if a file is being served
or a generic http proxy if the content is not a file

### QA Notes

On Server Web and Workbench, running the following in the corresponding
consoles:

##### Python

`pip install plotly nbformat pandas`

```python
import plotly.express as px
fig = px.bar(x=["a", "b", "c"], y=[1, 3, 2])
fig.show()
```

##### R

`install.packages('plotly')`

```r
library(plotly)
fig <- plot_ly(data = iris, x = ~Sepal.Length, y = ~Petal.Length)
fig
```

#### Expected Result

The corresponding interactive plots should display in the plots pane and
be interact-able!

timtmok committed:
Fix resizing plotly web clients (#4806)

Use Plotly's browser renderer
Intercept browser event to redirect to Plots view

midleman committed:
smoke tests: parallelization  (#4784)

### Intent

The goal is to optimize our E2E smoke tests, which currently run
serially, by parallelizing them to reduce overall execution time.

### Approach

Initially, all tests were run from a single file (`main.ts`), which
handled shared dependencies but prevented parallel execution. To address
this, I split the tests into multiple "main" files, which improved
execution times but made it challenging to balance workloads
effectively. Ultimately, the best solution was to refactor each test to
run independently, allowing for true parallelization.

Because `mocha-junit-reporter` doesn’t support parallel runs, I switched
to `mocha-xunit-reporter`, which natively handles parallel execution.
After the tests complete, I convert the results to the `junit` format to
ensure compatibility with TestRail.

### Impact on Workflow

1. ❌ Previously, when adding new test specs, we needed to manually
import and include them in the `main.ts` file. We no longer need to do
this! 🎉

2. ❌ We also no longer need to wrap all tests in `setup()`. 

3. ✅ The only thing we need to do is add `setupAndStartApp();`
(previously `installAllHandlers`) in the parent `describe` block of the
test
4. To tag tests for web, PRs, or windows, just add their respective tags
in the describe/it block (`#web`, `#win`, `#pr`)

### How to Run in Parallel

Use the `--parallel` flag and specify `--jobs`:

```bash
yarn smoketest-no-compile --tracing --parallel --jobs 2
```

### GHA Workflow Updates
I performed an analysis on run time for the full and pr test suites with
varying concurrent jobs. Summary of my findings:

**Full Test Suite** (avg time)

* **1 machine**:  ~25 minutes
* **2 machines**:  ~12 minutes
* **3 machines**:  ~9.5 minutes
  
**PR Test Suite** (avg time)
* **1 machine:**  ~6 minutes
* **2 machines:**  ~3 minutes 
* **3 machines:** ~2.5 minutes

To promote stability and prevent constrained resources, _I set GHA
workflows to use 2 jobs._ This is still a major speed improvement.

### QA Validation

* Confirmed that `--tracing` works as expected and all expected logs
exist.
* Verified that results (pass, fail, skip) are correctly imported into
TestRail. See
[example](https://posit.testrail.io/index.php?/runs/view/4353).
* Verified test report correctly displays on Workflow Summary

---------

Co-authored-by: Chris Mead <chris.mead@posit.co>

testlabauto committed:
Disable retries globally for smoke tests (#4872)

Retries might be creating very confusing traces, so disabling globally
for a windows workflow rerun.

### QA Notes

All smoke tests pass

jonvanausdeln committed:
Fixed layout issue for dataexplorer test on windows (#4873)

### Intent
Fixed failing data explorer test on windows CI

### Approach
R test was missing command to go into `notebook` layout

I also removed an extraneous #web identifier, as it was covered by the parent describe block.

### QA Notes

All tests in CI should pass

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

sharon-wang committed:
always send html plot content to a proxy server (#4876)

### Description

A follow-up to #4806 to revert
`src/vs/workbench/services/languageRuntime/common/languageRuntimeUiClient.ts`
back to using a proxy for interactive plots. A generic HTTP proxy was
added via #4855, so server plots can be proxied appropriately in Server
Web and Workbench.

Related to
- https://github.com/posit-dev/positron/issues/4804
- https://github.com/posit-dev/positron/issues/4274
- https://github.com/posit-dev/positron/issues/4245

### QA Notes

On Server Web and Workbench, running the following in the corresponding
consoles:

##### Python

`pip install plotly nbformat pandas`

```python
import plotly.express as px
fig = px.bar(x=["a", "b", "c"], y=[1, 3, 2])
fig.show()
```

##### R

`install.packages('plotly')`

```r
library(plotly)
fig <- plot_ly(data = iris, x = ~Sepal.Length, y = ~Petal.Length)
fig
```

#### Expected Result

The corresponding interactive plots should display in the plots pane and
be interact-able!

midleman committed:
readme: update smoke test section (#4877)

Updated the Smoke Test README to make it clearer and easier to follow.
Also added instructions for running tests in parallel locally.

jonvanausdeln committed:
Add beforeEach check for interpreter (#4881)

### Intent
Fix some instability in newproject tests where the interpreter startup intefers with typing in the project name.

### Approach
Allow inerpreter to finish loading in beforeEach for the suite of tests

### QA Notes
All smoke tests in CI should pass
<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

testlabauto committed:
Console output test: loop with short pauses (#4882)

Test for https://github.com/posit-dev/positron/pull/4848

Also, retries globally turned on for smoke tests with "double
screenshots and traces" as relevent.

### QA Notes

All smoke tests pass.

jmcphers committed:
Bump version to 2024.11.0 (#4884)

Bump Positron version to 2024.11.

jennybc committed:
Discover (non-current) R installations from Windows registry (#4878)

Addresses #4820

The goal is to discover non-current R installations from the Windows
registry. These were previously not discovered by Positron if the
installation also happened to be in a nonstandard location (and was not
on the PATH).

### QA Notes

The bare minimum would be to ensure that you current R installations
continue to be discovered, i.e. that this PR hasn't broken anything.
That is the most one could do on non-Windows.

More ambitious QA, which I have done, on Windows:

* Install R at a non-standard location. This will have to be via the
CRAN installer since rig doesn't give you any choice about this.
Therefore I removed my rig-installed released version of R and
re-installed released R with the CRAN installer below
`C:/nonstandardRLocation/`. Make sure the box to write to the registry
is checked (that is the default).
* Now use rig (or other means) to make *some other R version* the
current version. In my setup, R 4.2.3 and 4.3.3 are good candidates for
the new current version of R, e.g. `rig default 4.3.3`.
* Fire up Positron. You should see the released version of R, installed
at `C:/nonstandardRLocation/R-4.4.1` available to you in the interpreter
drop down after this PR. In a release build prior to this PR, that R
installation would not be picked up by Positron.

Other things that can be noticed in the Positron R output channel:

* Early on, you are likely to see the registry being consulted re: the
current R version:

  ```
2024-10-02 16:51:34.246 [info] Registry key
HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\InstallPath reports the current R
installation is at C:\Program Files\R\R-4.2.3
2024-10-02 16:51:34.246 [info] Identified the current R binary:
C:\Program Files\R\R-4.2.3\bin\x64\R.exe
  ```
This is happening when we validate metadata of the last-used R runtime.
The logging has been tightened up and this finding is also now cached,
which eliminates some repetitive discovery and logging that happens when
we do the main, broad search for R installations.
* These are completely new registry findings re: R installations that
may or may not be the current one:

  ```
2024-10-02 16:51:34.573 [info] Registry key
HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\4.2.3\InstallPath reports an R
installation at C:\Program Files\R\R-4.2.3
2024-10-02 16:51:34.573 [info] Registry key
HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\4.3.3\InstallPath reports an R
installation at C:\Program Files\R\R-4.3.3
2024-10-02 16:51:34.573 [info] Registry key
HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\4.4.1\InstallPath reports an R
installation at C:\nonstandardRLocation\R-4.4.1
2024-10-02 16:51:34.573 [info] Registry key
HKEY_LOCAL_MACHINE\SOFTWARE\R-core\R64\4.5.0 Pre-release\InstallPath
reports an R installation at C:\Program Files\R\R-devel
  ```
The installations below `C:\Program Files\R` would be discovered anyway,
because they are in a well-known place for this OS. What's new is that
we find `C:\nonstandardRLocation\R-4.4.1`.
* We no longer message about registry keys that are not found, because
folks were regularly misinterpreting that as some sort of error, whereas
it's expected that we might check for keys that aren't defined.

isabelizimm committed:
hopefully fixing Buffer problems???

midleman committed:
smoke tests: add ability to target build (#4880)

### Intent
Restructure the Mocha test runner code to ensure that all test setup and
environment preparation steps are executed before the actual test suite
begins. This restructuring was necessary to allow the test runner to
properly target a specific Positron build using the `--build` flag.

### Approach
* Moved all environment setup and configuration logic to run before the
test suite starts.
* Refactored code to handle `--build` path validation and setup early in
the execution.
* Ensured environment variables are correctly defined during the setup
phase.

### QA Notes
To target a build:
```
yarn smoketest-pr --build /Applications/Positron.app
```

The console will log the version of the targeted build:
<img width="702" alt="Screenshot 2024-10-03 at 9 39 49 AM"
src="https://github.com/user-attachments/assets/47b6d127-2f70-496b-b275-d209de40eea4">

* Ran full test suite against PR and things look good. 👍

testlabauto reviewed (dismissed):
Wow.  That is huge.  I scanned through it and am impressed.  I see a lot of changes that I can easily follow and feel comfortable with and a fair amount that are beyond my skills, but my feeling is that if tests are passing lets merge and continue to test further.

sharon-wang commented on extensions/positron-python/src/test/positron/webAppCommands.unit.test.ts:
```diff
@@ -180,9 +182,12 @@ suite('Web app commands', () => {
         isFastAPICliInstalled = false;
 
-        sinon.stub(vscode, 'workspace').get(() => ({
+        // TODO @isabelizimm: HACK HACK HACK
+        const mockAsRelativePath = sinon.mock().callsFake(() => ({
             workspaceFolders: [{ uri: { fsPath: '/path/to' } }],
         }));
+        // Replace the original `asRelativePath` method with the mock object
+        vscode.workspace.asRelativePath = mockAsRelativePath;
```
I wonder if we can omit this entirely? The other FastAPI tests don't seem to have this stub and if I'm understanding this test correctly, I think the main aspect of this test is passing `undefined` to `verifyRunAppCommand` so the app name cannot be inferred

```suggestion
```

testlabauto committed:
Adjust data explorer case timeouts (#4896)

Reliability fix - give select date explorer cases longer to finish

### QA Notes

All smoke tests should pass.

midleman committed:
smoke tests: organize (#4894)

### Intent
Now that test parallelization is in place and working (🤞), I revisited
the code to improve its organization and maintainability.

### Approach
* All helpers used to run the tests are now contained in
`test/smoke/test-runner`.
* The actual script that kicks off the tests is now:
`test/smoke/run-tests.js`

### QA Notes
* Confirmed full test suite and windows suite ran w/o issue.
* Confirmed logs are still there.
* Oh and turned retries back on.

DavisVaughan committed:
Bump ark to 0.1.143 (#4895)

Ensure we get green lights across the board here first
https://github.com/posit-dev/ark/actions/runs/11182113440

- https://github.com/posit-dev/ark/pull/536

midleman committed:
smoke tests: remove declarations (#4897)

Remove generate declarations for `test/smoke` dir.

isabelizimm commented on extensions/positron-python/python_files/positron/pinned-test-requirements.txt:
```diff
@@ -1,24 +1,27 @@
 bokeh==3.4.3; python_version >= '3.9'
```
We have some info in https://github.com/posit-dev/positron-wiki/blob/main/fork-management.qmd, but not much on how to update this file. 

I mostly copied the last-known working package version for Python versions that are no longer supported, bumped the rest via these instructions, and ran CI to see what broke--I'm not sure of a programmatic way to do easily do this beyond making a different env per Python version we support and then upgrading -> freezing requirements.txt files and then concatenating them?

Looking at the [Nightly CI where we get the latest version of a package](https://github.com/posit-dev/positron/actions/workflows/positron-python-nightly.yml) and comparing versions might be a less painful manual way.

sharon-wang commented on package.json:
```diff
@@ -241,7 +241,7 @@
     "tsec": "0.2.7",
     "tslib": "^2.6.3",
-    "typescript": "^5.7.0-dev.20240826",
+    "typescript": "^5.7.0-dev.20240827",
```
Yeah, it seems like the root package.json version of typescript that is compatible with this python extension update has deviated since our last upstream merge of vscode, so now we need `20240927`

seeM replied:
I'm nervous about changing the root TS version based on one extension. Could it lead to issues with other built-in extensions? Is it possible to rather pin the Python version to match the root version?

isabelizimm replied:
I first believed that matching the python extension to root would break the build in the same way as we were seeing in [this CI run](https://github.com/posit-dev/positron/actions/runs/11117084279/job/30888442108?pr=4670), but it seems to build locally. Pushed the change and trying it out now 👀

isabelizimm committed:
Update extensions/positron-python/src/test/positron/webAppCommands.unit.test.ts

Co-authored-by: sharon <sharon-wang@users.noreply.github.com>
Signed-off-by: Isabel Zimmerman <54685329+isabelizimm@users.noreply.github.com>

isabelizimm committed:
Update package.json

Co-authored-by: sharon <sharon-wang@users.noreply.github.com>
Signed-off-by: Isabel Zimmerman <54685329+isabelizimm@users.noreply.github.com>

midleman committed:
smoke tests: fix env vars when targeting builds (#4899)

Not totally sure why, but _only_ when targeting builds the env vars were
not properly being set. To ensure that they are manually setting them in
the test-hooks file.

midleman committed:
smoke tests: add `only` script (#4900)

Per conversation with @testlabauto , it might be beneficial to have
ability to run a subset of tests in parallel locally. Now we can just by
tagging tests with `#only` in test description.

### QA notes
* updated README to reflect these changes
* confirmed with `mocah.dryRun()` that web, pr, all still run the
expected the of filtered tests.

jonvanausdeln committed:
Skip R console input test (#4902)

Skipping R console input test due to
https://github.com/posit-dev/positron/issues/4901

isabelizimm committed:
ruff

isabelizimm committed:
remove ! as any

isabelizimm committed:
add in a few more ruff rules

isabelizimm commented on extensions/positron-python/python_files/create_venv.py:
```diff
@@ -89,9 +89,9 @@ def venv_exists(name: str) -> bool:
-        subprocess.run(args, cwd=os.getcwd(), check=True)
-    except subprocess.CalledProcessError:
-        raise VenvError(error_message)
+        subprocess.run(args, cwd=os.getcwd(), check=True)  # noqa: PTH109
```
They added LOTS of ruff rules upstream, which I did not pull in since I'd rather have it be a follow up PR due to the amount of code changes (800+ errors that are not auto-fixable). I'll update the `pyproject.toml` to match upstream with the breaking rules commented out to pass CI.

testlabauto committed:
Cmead/reliability fixes oct4 (#4903)

A couple test reliability fixes:
* retry more intelligently for failed filter setting
* increased timeouts

### QA Notes

All smoke tests passing

isabelizimm committed:
fixes from review

isabelizimm committed:
nit

isabelizimm committed:
nit to pyproject.toml

midleman committed:
r-markdown smoke test: bump timeout (#4906)

### QA Notes

Bumping timeout on the R-markdown test.

DavisVaughan committed:
Only hide editor if panel hasn't been explicitly hidden (#4908)

Addresses https://github.com/posit-dev/positron/issues/4756

This has been driving me _crazy_! I am one of those people that prefers
to move everything out of the Panel and then _explicitly hide it_ with
`CMD + J`.

Due to https://github.com/posit-dev/positron/pull/1782/ at this request
https://github.com/posit-dev/positron/issues/980, when:
- The last open editor closes
- Or when there are no editors open on startup

then the Panel is forcibly popped up and maximized. This is useful if
your Console lives there, but is super annoying if you don't have
anything in the Panel and have it "hidden" (see the video in
https://github.com/posit-dev/positron/issues/4756).

I think that if we also require that the Panel be "visible" (i.e. the
user hasn't explicitly hidden it) before hiding the editor, then we can
respect the spirit of the request in
https://github.com/posit-dev/positron/issues/980 while also respecting
people like me who hide the Panel.

Video shows:
- Panel stays hidden after last tab is closed
- Panel stays hidden after a reload (simulating opening Positron)
- Then we make the Panel visible and try those again, to prove that the
old behavior is still in place


https://github.com/user-attachments/assets/8e3cca0f-4bb1-48ef-acde-ae017c49c7cc

midleman committed:
notebookCreate test: fix test setup (#4922)

### Intent
Noticed two issues with this test. First, the tests within each describe
block are not independent and rely on the first one. So I re-organized
the before/beforeEach blocks so each test can standalone. The 2nd issue
is that if the 2nd tests attempt to retry they will potentially fail
because there will be duplicate matching markdown text from the first
attempt.

### Approach
1. New notebook for each test
2. Random text used to verify h2 markdown

jonvanausdeln committed:
added platform specific paths to smoke test --build flag (#4911)

### Intent

The `--build` flag failed on windows & linux because the getPositronVersion only used MacOS pathing.

### Approach
Added correct pathing for each OS

## QA Notes
Using the --build flag works on all platforms

midleman committed:
better caching (#4923)

### Intent
I previously implemented caching but had to partially [roll it
back](https://github.com/posit-dev/positron/pull/4771/files) because I
mistakenly tied the nested directories to the root lock files for
`extensions` and `build`. My goal now is to correct this setup and
restore caching for the nested directories.

### Approach
I separated the `extensions` and "extensions nested" directories into
distinct caches. For the nested caches, I’m using a hash generated from
**all** the existing yarn.lock files in the nested directories. This
ensures that if a single extension changes, the cache is updated
accordingly. I applied the same approach to the `remote` directory and
its nested extensions. For example:
```bash
    - name: Cache nested extensions
      id: cache-nested-extensions
      uses: actions/cache@v4
      with:
        path: ./extensions/**/node_modules
        key: cache-nested-extensions-v2-${{ runner.os }}-${{ hashFiles('extensions/**/yarn.lock') }}
```
In this example, the cache key is built using the content of all
`yarn.lock` files under `extensions/**/yarn.lock`. If there are separate
`yarn.lock` files in the root extensions directory and in its
subdirectories, each will be hashed independently, making the root and
nested caches distinct from one another.

### QA Notes
* I [temporarily modified a nested lock
file](https://github.com/posit-dev/positron/pull/4923/commits/1ff42523c26cce3408971f2e1c75718a773bd60b)
and confirmed the caching behaved as expected.
<img width="910" alt="Screenshot 2024-10-07 at 1 21 25 PM"
src="https://github.com/user-attachments/assets/cebd80ad-007d-49a0-9f0a-946a9d5c4d6c">

testlabauto committed:
Enable more web plot tests; fix data explorer filter workaround (#4927)

Enable the currently functioning web tests which required special
handling for R webview tests.

Fix an oversight in the error/retry handling for adding a data frame
filter.

### QA Notes

All smoke tests should pass

jonvanausdeln committed:
Upload windows test results (#4931)

### Intent

Addresses #4866

Adds slack and TestRail reporting for windows nightly run


### QA Notes
Verified pass, fail & skipped results are uplaoded to Testral
Verified failed run reports to slack

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

sharon-wang committed:
default to HTML file proxy unless uri is http or https (#4939)

- Addresses https://github.com/posit-dev/positron/issues/4930

On Windows, a file path has scheme `C`, which was not handled in
https://github.com/posit-dev/positron/pull/4855. I had assumed local
file paths would all have scheme `file`, but that is not the case on
Windows!

I've updated the logic to start a generic HTTP proxy server when the
scheme is `http` or `https`, but to assume all other paths are HTML file
paths.

The behaviour of `startHtmlProxyServer` before
https://github.com/posit-dev/positron/pull/4855 was to always start an
HTML proxy server, so this PR restores that assumption by always
starting an HTML proxy server unless we have an `http`/`https` URL (in
which case we want to start a generic proxy server).

### QA Notes

See #4930 for repro notes.

Co-authored-by: sharon wang <25834218+sharon-wang@users.noreply.github.com>

timtmok committed:
Add Open Folder to web Welcome (#4929)

Enables the `Open Folder...` action for the Welcome page

jonvanausdeln committed:
Added caching to windows workflow (#4943)

Added caching step to the workflow

### QA Notes
All tests pass
<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

jmcphers committed:
Add new Jupyter kernel supervisor (#4910)

This change adds a new and currently optional Jupyter kernel supervisor
to Positron.

When enabled, all R and Python kernels are run under a supervisor
process (codenamed 'Kallichore'). This process is a native binary that
takes care of the minutiae of kernel lifecyle management (process
output, startup, shutdown, restart, status reporting, etc.). It also
takes care of the low-level ZeroMQ socket connectivity and exposes a
simple WebSocket based interface that proxies messages to and from the
ZeroMQ sockets.

```mermaid
graph TD
R[Positron R Extension] -- Jupyter Adapter API --> ka[Kallichore Adapter]
ka -- WebSocket --> k[Kallichore Supervisor]
k -- ZeroMQ --> ark[R Jupyter Kernel - ark]
py[Positron Python Extension] -- Jupyter Adapter API --> ka
k -- ZeroMQ --> ipy[Python Jupyter Kernel - ipykernel]
```

The implementation is a drop-in replacement for the Jupyter Adapter and
currently exposes an API that is fully compatible with the Jupyter
Adapter; the only changes to the R and Python extensions are switches
that determine which extension the Jupyter Adapter API is loaded from.
The change is otherwise totally transparent to the R and Python
extensions since they already are behind a layer of abstraction from the
Jupyter protocol.

The current plan of record is for the Jupyter Adapter to be fully
deprecated in favor of this new supervisor when it reaches feature
parity, stability, and performance goals. This will allow us to get rid
of our troublesome ZeroMQ bindings in Node and will also improve
stability and performance around session management.

The goal of this PR is to start exposing the new supervisor to builds
and to make it possible to start collaborating on work associated with
the supervisor. While most everything works today with the new
supervisor, there are a lot of edge cases that aren't finished yet, so
the supervisor is disabled by default and will continue to be until it's
ready to take over.

Lint-level issues are aggregated here:
https://github.com/posit-dev/positron/issues/4912

This change isn't ready for QA yet. It tries very hard to avoid touching
any code outside the new adapter extension; the only thing worth testing
is probably R and Python restarts.

jonvanausdeln committed:
Add Plots test to Windows Workflow (#4951)

### Intent

Part of addressing #4867 adds plots tests

### Approach

* Had to modify some to expand various panes to be visible
* Had to modify the `simplePlotTest` to make pane visible
* Plots that compare to known good are skipped because the windows diff are too great. we could later add some OS specific base images, but ROI not worth it now.
* Save plots skipped because the save dialog overflows the screen.

### QA Notes

Windows & Linux Nightly runs works as expected.

seeM committed:
Fix saving new notebooks reverting to Python (#4933)

Addresses https://github.com/posit-dev/positron/issues/4796.

seeM committed:
Support debugging Python web applications (#4924)

This PR adds "Debug \<Framework\> App in Terminal" commands for all of
the currently supported Python app frameworks.

There was a fair bit of moving things around to reuse code between the
run and debug commands.

Addresses #4795.

### QA Notes

Test the commands for each app framework, see the examples for each
framework in #4662.

testlabauto committed:
Python App Tests (#4947)

Python App Tests
* Dash
* Gradio
* FastApi - disabled and bug filed for ubuntu
* Streamlit

Only streamlit currently works is web mode, so web execution is only
enabled for it.

Note that you may need to update your streamlit package to get this to
pass locally.

### QA Notes

All smoke tests should pass

timtmok committed:
Fix plotly showing in Viewer (#4904)

seeM committed:
Fix switching cell type loses language (#4934)

Addresses https://github.com/posit-dev/positron/issues/4898.

I don't think the previous note regarding not setting
`supportedLanguages` is an issue since we implemented automatic kernel
selection.

testlabauto committed:
Fix for flaky interpreter dropdown test and more logging (#4958)

Allow more time for evaluating the state of the interpreter dropdown
post restart and log specific failure when error occurs.

Also, minor cleanup of import in pythonApps.test.ts.

### QA Notes

All smoke tests should pass.

jennybc committed:
Don't leak env vars across tasks (#4954)

Addresses #4953

jonvanausdeln committed:
Update filenames used from qa-example-content repo (#4969)

<!-- Thank you for submitting a pull request.
If this is your first pull request you can find information about
contributing here:
  * https://github.com/posit-dev/positron/blob/main/CONTRIBUTING.md

We recommend synchronizing your branch with the latest changes in the
main branch by either pulling or rebasing.
-->

<!--
  Describe briefly what problem this pull request resolves, or what
  new feature it introduces. Include screenshots of any new or altered
  UI. Link to any GitHub issues but avoid "magic" keywords that will 
  automatically close the issue. If there are any details about your 
approach that are unintuitive or you want to draw attention to, please
  describe them here.
-->

### QA Notes

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

testlabauto committed:
Bump python app test timeouts up (#4975)

Bump python app test timeouts to 45 seconds. Sometimes these were
failing in local parallel runs.

### QA Notes

All smoke tests should pass.

testlabauto committed:
Skip vetiver test (#4980)

Skipping test based on an open issue.

### QA Notes

All smoke tests pass.

jonvanausdeln committed:
Add Python App tests to windows workflow (#4981)

seeM committed:
Fix notebook stream output ANSI code handling (#4937)

Addresses #3723. The main change is in `handleRuntimeMessageStream`,
where we now append to an existing output if possible.

seeM committed:
Support prompts in notebooks (#4938)

Addresses https://github.com/posit-dev/positron/issues/4199.

timtmok committed:
Urls always opening in Plots view (#4966)

jmcphers committed:
Allow workspace override for interpreter settings (#4974)

This change allows interpreter-related settings to be configured at the
workspace level, if desired.

midleman committed:
cicd: parallelize full smoke test (#4950)

### Intent

The goal here is to speed up our Full Test Suite by running all tests
(integration, unit, and e2e) in parallel, instead of serially.

### Approach

Each test now has its own runner, so they can all run all at once
instead of waiting on each other. For now, the build and compile steps
are still happening in each test job, but in the future, we could
optimize this further by moving those steps to an upstream job and
sharing the built artifacts with the test jobs that follow.

#### Previously ~ 54min
<img width="627" alt="Screenshot 2024-10-10 at 8 23 43 AM"
src="https://github.com/user-attachments/assets/986fd3e5-e126-4848-b6ea-fe12428208ed">

#### After changes ~ 27min
<img width="729" alt="Screenshot 2024-10-10 at 8 22 37 AM"
src="https://github.com/user-attachments/assets/ca3d83d3-75dc-4073-a96e-2fd39fbec03e">


### QA Notes
* Confirmed PR tests pass
* Confirmed full test suite tests pass
* I did not test slack notifications, but I made NO changes to the
logic/behavior.
* Kept running into failures/flakes from the interpreter dropdown test,
so skipped it for now and will look into after this is merged.

testlabauto committed:
Swap skipping for Python viewer tests (#4989)

Vetiver works now, but great-tables will randomly fail, only in CI.
Disabling for the moment

### QA Notes

All smoke tests should pass.

midleman committed:
cicd: license tweaks on full test suite (#4992)

midleman committed:
mocha runner: better logs around test cleanup (#4993)

### Intent
When running a large number of tests in parallel, the script cleans up
the test data generated during the run, which can sometimes take a
while. To save time, there’s an option to skip this step, commonly used
in CICD pipelines. This option was already available, but the recent
update improves its visibility to make it clearer to the person running
the tests.
Note: This does not delete traces, but rather temporary files created by
the runner.

### Approach
Enhanced the logging to provide better visibility, highlighting when the
cleanup is being skipped to reduce confusion and improve monitoring.

#### Without skipping clean up
<img width="583" alt="Screenshot 2024-10-10 at 2 20 32 PM"
src="https://github.com/user-attachments/assets/5522e993-1f6d-4e16-8ec1-0d18ad654e76">

#### With skipping clean up  (`--skip-cleanup`)
<img width="220" alt="Screenshot 2024-10-10 at 2 20 02 PM"
src="https://github.com/user-attachments/assets/5b909df4-535f-433b-9c3a-390953fe7a59">

midleman committed:
cicd reporter: remove unnecessary dep (#4999)

Removing a dependency that is not needed/used.

### QA Notes

Confirmed reports still generate w/o issue.

midleman committed:
cicd: move xvfb config into right composite action (#4995)

Just moving the `xvfb` config into the correct location.

Running ~[full
suite](https://github.com/posit-dev/positron/actions/runs/11281852588)~
~[this
one](https://github.com/posit-dev/positron/actions/runs/11282126825)~
[for real, this
one](https://github.com/posit-dev/positron/actions/runs/11290802008) to
be sure. ;)

jonvanausdeln committed:
Add Shiny, Connections, and console tests to windows (#5004)

### Intent
Continuing #4867, added 
* Shiny
* Connecctions
* Remaining Console

### Approach

Only consoleHistory required test modification to add hiding the primary side bar.

### QA Notes

All test pass in both Linux and WIndows runs.

<!--
  Add additional information for QA on how to validate the change,
  paying special attention to the level of risk, adjacent areas that
  could be affected by the change, and any important contextual
  information not present in the linked issues.
-->

testlabauto committed:
Run smoke tests against latest release (#4990)

Run the current branch's smoke tests against the latest release

Requires a build of Positron, though.

### QA Notes

Latest release is used & smoke tests pass.

midleman committed:
e2e tests: add retry.log (#5005)

### Intent

We are currently logging stack traces in the test report, which makes
the logs difficult to read and negatively impacts the xUnit/JUnit
reports. To retain this important information without cluttering the
reports, I moved the stack traces to a `retry.log` file within each
test’s directory.

### Approach

Each test already has its own directory for relevant logs. I added a
`retry.log` file to this directory, which captures the stack traces for
all retry attempts of that specific test.

<img width="281" alt="Screenshot 2024-10-11 at 2 13 58 PM"
src="https://github.com/user-attachments/assets/908250ca-2dfd-4fed-8018-f12c6e5ef0a7">
<img width="806" alt="Screenshot 2024-10-11 at 2 13 48 PM"
src="https://github.com/user-attachments/assets/7418233d-6742-4a57-bc7b-0c1f7768f6f8">

### QA Notes
Confirmed that all reports and traces are still present.

sharon-wang committed:
Proxy python app frameworks (#4978)

## Description

- Addresses https://github.com/posit-dev/positron/issues/4769 (as much
as possible for now)

### Framework Support Summary (based on **_limited_** testing)
| Framework | Positron Server Web | Positron Desktop | Positron on
Workbench | Notes |
|--------|--------|--------|--------|--------|
| Dash | ✅ yes | ✅ yes | ✅ yes-ish | [Workbench extension
issue](https://github.com/rstudio/rstudio-workbench-vscode-ext/issues/262)
to skip Dash framework handling to avoid conflicting with Positron's
Dash framework handling |
| Fastapi | ✅ yes | ✅ yes | 🛑 no | Seems to be a conflict between the
way we run a fastapi app and that Workbench is setting
`UVICORN_ROOT_PATH` |
| Flask | ✅ yes | ✅ yes | ✅ yes | Workbench: works when including the
code snippet from [Workbench
docs](https://docs.posit.co/ide/server-pro/user/vs-code/guide/proxying-web-servers.html#flask).
|
| Gradio | ✅ yes-ish | ✅ yes-ish | ✅ yes-ish | Working when the
following dependency versions are used: `gradio==3.3.1 fastapi==0.85.2
httpx==0.24.1`. See https://github.com/gradio-app/gradio/issues/9529 for
more information on why more recent versions don't work. |
| Streamlit | ✅ yes | ✅ yes | 🟨 partially | Not working on Workbench
when SSL is enabled |
| Shiny | ✅ yes | ✅ yes | ✅ yes | | 

### Implementation Notes

#### Positron Proxy
CC @softwarenerd & @jmcphers
- Add new command `positronProxy.startPendingProxyServer` which starts a
new proxy server, but doesn't set up the middleware right away. The
command will return the `externalUri` of the proxy server (this is the
same as the url shown in the Viewer), the `proxyPath` (which some app
frameworks use as the `urlPrefix`) and `finishProxySetup()` which the
command-caller will invoke, passing the `targetOrigin` (the actual app
url) so that we can finish setting up the middleware to proxy requests
from the `externalUri` (app proxy uri) to the `targetOrigin` (actual app
uri).
- Refactor the proxy code so we can set up a proxy in multiple steps,
not only all-at-once
- existing callers of `startProxyServer()` will continue to use that
method which still does the all-at-once proxy setup, by calling
`startNewProxyServer()` and `finishProxySetup()` in succession
- new option to call `startNewProxyServer()` standalone, which will only
start up the server and then call `finishProxySetup()` later, once the
`targetOrigin` is known
- Move the HTML-rewriting code to the util file

#### Positron Run App
CC @seeM
- expand the `localUrlRegex` to include the path after the url port
`<HOST>:<PORT>/<PATH>`
- execute the command `positronProxy.startPendingProxyServer` before
setting up terminal options or the debug configuration, so we can start
the proxy server and get the `urlPrefix`
- before previewing the url in the Viewer, finish the proxy setup by
passing the `localUri` to the positron proxy via
`proxyInfo.finishProxySetup(localUri.toString())`, so that the proxy
middleware is set up
- remove `port` from types and test files as it is unused
- increase timeout for waiting for the app url to show in the terminal
(in particular, Gradio would take a bit longer on Workbench and we would
timeout)

#### Python Extension
CC @seeM 
- remove `port` from the `webAppCommands` and related test files which
is unused
- update the framework-specific app arguments and environment variables
to work with our proxy server situation

#### Custom `resolveExternalUri`
CC @melissa-barca
- update the `resolvedUri` to inherit the protocol used by the main
window as the uri's scheme, so we can upgrade to `https` if the main
window is running in a secure context
- NOTE: this will need to be upstreamed

### QA Notes

This PR involves refactoring the positron-proxy, which is used by Help,
Interactive Plots and the Viewer. We should not see any regression in
these types of proxied content.

Positron Server Web and Desktop should be working across all app
frameworks in:
-
https://github.com/posit-dev/qa-example-content/tree/main/workspaces/python_apps
- When running the Gradio app, you will need to install these versions
`gradio==3.3.1 fastapi==0.85.2 httpx==0.24.1`. If you can get it working
with a more recent combination of versions, please let me know!
-
https://github.com/posit-dev/qa-example-content/tree/main/workspaces/shiny-py-example
-
https://github.com/posit-dev/qa-example-content/tree/main/workspaces/streamlit-py-example

Positron on Workbench:
- `dash` works if the generated `.env` file is deleted before running.
Once [this
issue](https://github.com/rstudio/rstudio-workbench-vscode-ext/issues/262)
is complete, this extra step won't be needed.
- `fastapi` does not work. The current hunch is that Workbench setting
`UVICORN_ROOT_PATH` at session start is interfering with how we run the
`fastapi` app with `uvicorn`. The `UVICORN_ROOT_PATH` is set based on
the default fastapi port `8000`, however I think we want the proxied app
url as the root path, which is provided by the Positron Proxy. But, when
setting `--root-path` in the fastapi app command to the proxied root
path, it does not work in Server Web or Desktop. TBD if this resolves
the issue on Workbench. More investigation needed.
- `flask` works as long as you include the code snippet from [Workbench
docs](https://docs.posit.co/ide/server-pro/user/vs-code/guide/proxying-web-servers.html#flask).
- `gradio` works when these versions are installed `gradio==3.3.1
fastapi==0.85.2 httpx==0.24.1`. There's [an
issue](https://github.com/gradio-app/gradio/issues/9529) in newer
versions of Gradio when the app is run via a proxy.
- `streamlit` does not work when SSL is set up. More investigation
needed.
- `shiny` no notes!

---------

Co-authored-by: sharon wang <25834218+sharon-wang@users.noreply.github.com>

isabelizimm commented:
this has honestly been such a nightmare due to config issues, nothing even relating to changes 😩 if I get green CI, I'll merge for further testing

isabelizimm committed:
updates from linting

isabelizimm commented:
everything is green and dandy but git gymnastics are bogging me down-- i'm going to cut my losses here and hop to a fresh branch
